### DISCUSSION TITLE
Provide username (and password) for authentication request

### DISCUSSION THREAD

**Discussion by Phil0L** [spotify-web-api-java/spotify-web-api-java#330] [Q&A] [answered]
Hello everybody,

I'm trying to create an application that can control multiple Spotify accounts.
For this I make multiple authentication requests via authentication code auth flow.
It works fine and I can get multiple access tokens.
The only problem I have is that the api seems to only remember the last used credentials.
What's currently happening:
The application requests access token 1 for user 1 -> I log in and grant all permissions.
The application requests access token 2 for user 2 -> I log in and grant all permissions.
Restart of the application.
The application requests access token 1 for user 2 ↯

What I'm trying to achieve:
Upon restarting the application, all access tokens should be received without user input again

I'm thinking of two possibilities, but I can't seem to find a way to achieve them.
Either save not only the last user, but several
Or provide the username (+ password) in the authentication request.

Anyone have any idea how to do something like this?
Thanks in advance

**Comment by dargmuesli** [comment, marked as answer]
How do you persist access tokens? Sounds like a code example might be of use here.

**Reply by Phil0L** [reply]
I don't.
They are just stored in a variable.
When I restart the application, they are lost.

This is how my authentication works atm:

`
    
    private static final String[] SCOPES = new String[]
            {"user-read-private", "streaming", "user-top-read", "user-read-recently-played",
                    "app-remote-control", "user-read-playback-state", "user-modify-playback-state",
                    "user-read-currently-playing", "user-read-playback-position", "user-read-email"};
    private final String clientID;
    private final String clientSecret;
    private final String redirectUri;
    private final String username;
    private final String password;
    private final SpotifyApi api;

    private AuthenticationCallback callback;

    public SpotifyAuthHelper(String clientID, String clientSecret, String redirectUri, String username, String password) {
        this.clientID = clientID;
        this.clientSecret = clientSecret;
        this.redirectUri = redirectUri;
        this.username = username;
        this.password = password;
        this.api = SpotifyApi.builder()
                .setClientId(clientID)
                .setClientSecret(clientSecret)
                .setRedirectUri(SpotifyHttpManager.makeUri(redirectUri))
                //.setHttpManager(new CustomHttpManager(username, password))
                .build();
    }

    public void authenticate(AuthenticationCallback callback) {
        this.callback = callback;
        URI uri = api.authorizationCodeUri().scope(String.join(",", SCOPES)).build().execute();
        System.out.println("Auth url: " + uri.toString());
    }

    private void onCode(String code) {
        AuthorizationCodeCredentials credentials = api.authorizationCode(code).build().executeAsync().join();
        api.setAccessToken(credentials.getAccessToken());
        api.setRefreshToken(credentials.getRefreshToken());
        connected();
    }

    private void connected() {
        if (callback != null)
            callback.onAuthentication(api);
    }
`

And then the execution:

`
    private final SpotifyAuthHelper authHelper1 = new SpotifyAuthHelper(CLIENT_ID1, SECRET_1, REDIRECT_URI1, USERNAME1, PASSWORD1);
    private final SpotifyAuthHelper authHelper2 = new SpotifyAuthHelper(CLIENT_ID2, SECRET_2, REDIRECT_URI2, USERNAME2, PASSWORD2);
    private SpotifyApi api1;
    private SpotifyApi api2;

    private void authenticate1() {
        authHelper1.authenticate(api -> {
            this.api1 = api;
            authenticate2();
        });
    }

    private void authenticate2() {
        authHelper2.setAndInitWebView(mWebView);
        authHelper2.authenticate(api -> {
            this.api2 = api;
            connected();
        });
    }
   
`

**Reply by dargmuesli** [reply]
It's not clear to me how 

> the api seems to only remember the last used credentials

works while

> When I restart the application, they are lost

and you 

> Restart of the application

Why would

> The application requests access token 1 for user 2 ↯

If

> They are lost

**Reply by Phil0L** [reply]
If I just take a spotify account and start the authentication process, then spotify wants my access data and after entering it redirects me to the redirect url. now when i stop and restart the application, spotify redirects me to the redirect url without any prompt. this is the behavior i want. I want you to only have to log in once. so how does the api (or spotify or whoever) store those last used credentials? And do I get it somehow that 2 credentials are saved? Or that I send the credentials via api?

**Reply by dargmuesli** [reply]
This api wrapper does not store anything, the REST api itself should neither. It seems like it's up to you to set a breakpoint and step through your code to see why a url is opened.
You're correctly using 2 api instances as far as I can see.

**Reply by Phil0L** [reply]
So...
First of all, thank you very much for your help.
I was very stupid. I overlooked the one obvious thing. The REFRESH TOKEN...
I now persist the refresh token in a file, to recreate an access token at any time.
The refresh token does not expire right?

I will mark this discussion as answered.

**Reply by dargmuesli** [reply]
I'm not sure, actually. I think some but not all authorization flow implementations (in the www) do set a (very long) expiry for refresh tokens. Not sure about Spotify.
